segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/stat/AutoScaleProcessor.java

```diff
@@ -116,48 +123,51 @@
                 }, executor))
                 .build();
 
+        this.executor = executor;
+        
         // Even if there is no activity, keep cleaning up the cache so that scale down can be triggered.
         // caches do not perform clean up if there is no activity. This is because they do not maintain their
         // own background thread.
         this.cacheCleanup = executor.scheduleAtFixedRate(cache::cleanUp, 0, configuration.getCacheCleanup().getSeconds(), TimeUnit.SECONDS);
-        if (clientFactory != null) {
-            bootstrapRequestWriters(clientFactory, executor);
-        }
     }
 
     @Override
+    @Synchronized
     public void close() {
-        val w = this.writer.get();
-        if (w != null) {
-            w.close();
-            this.writer.set(null);
+        if (writer != null) {
+            writer.cancel(true);
+
+            if (Futures.isSuccessful(writer)) {
+                val w = this.writer.join();
+                if (w != null) {
+                    w.close();
+                }
+            }
         }
 
         this.clientFactory.close();
         this.cacheCleanup.cancel(true);
     }
 
-    private void bootstrapRequestWriters(EventStreamClientFactory clientFactory, ScheduledExecutorService executor) {
-        // Starting with initial delay, in case request stream has not been created, to give it time to start
-        // However, we have this wrapped in consumeFailure which means the creation of writer will be retried.
-        // We are introducing a delay to avoid exceptions in the log in case creation of writer is attempted before
-        // creation of requeststream.
-        executor.schedule(
-                () -> Retry.indefinitelyWithExpBackoff(100, 10, 10000, this::handleBootstrapException)
-                        .runInExecutor(() -> bootstrapOnce(clientFactory), executor),
-                10, TimeUnit.SECONDS);
+    @Synchronized
+    private CompletableFuture<EventStreamWriter<AutoScaleEvent>> getWriter() {
+        if (writer == null) {
+            AtomicReference<EventStreamWriter<AutoScaleEvent>> w = new AtomicReference<>();
+            EventWriterConfig writerConfig = EventWriterConfig.builder().build();
+            writer = Retry.indefinitelyWithExpBackoff(100, 10, 10000, this::handleBootstrapException)
```

Comment:
This indefinite retry won't work well in case we fail to bootstrap and we decide to shut down. It will likely leave this running forever. You should include a condition that can cancel it. My recommendation would be to check `this.writer.isDone` (it will be true when either a writer has been set or when it got completed with an exception).